### PR TITLE
Prisms NC AMR support [nc-prism-dev]

### PR DESCRIPTION
**Pull request by jakubcerveny**
Prism support in `NCMesh`. This mostly entails support for triangular faces and their traversal. Once this works, adding non-conforming tets should be easy (if needed).

TODO:
- [x] There still seems to be a bug in the P matrix construction, need to debug.
- [x] New situation in forced refinement propagation. Edge neighbor to an aniso face (middle wedge) currently not handled (a little tricky, didn't expect that).
- [x] Coarse-fine transforms for wedges (easy).
- [x] Wedge derefinement.
- [x] LimitNCLevel for wedges
- [x] Parallel P matrix: triangle faces on CPU interfaces
- [x] Fix problem with master edge being constrained by a face
- [x] Fix rebalance bug (probably caused by neighbor calculation)
- [x] Update CHANGELOG

### CODE REVIEW

**Comment by mlstowell** [comment]
It was a very pleasant surprise to see this PR.  Thanks @jakubcerveny!

**Comment by jakubcerveny** [comment]
Example 15 (serial) works for prisms: 

ex15-b

**Comment by jakubcerveny** [comment]
Unfortunately `ex15p` won't work easily because we don't have `RT_WedgeElement` (needed in `L2ZienkiewiczZhuEstimator`).

**Comment by tzanio** [comment]
> Unfortunately `ex15p` won't work easily because we don't have `RT_WedgeElement` (needed in `L2ZienkiewiczZhuEstimator`).

@mlstowell is working on it I believe :)

**Comment by tzanio** [comment]
We can also switch to a different error estimator...

**Comment by v-dobrev** [comment]
You can try replacing the `L2ZienkiewiczZhuEstimator` with `ZienkiewiczZhuEstimator` (which now should work in parallel) as in the serial `ex15`.

**Comment by jakubcerveny** [comment]
Aniso prism refinement finally works:
aniso-prisms

**Comment by jakubcerveny** [comment]
`ex15p` works for prisms (with `ZienkiewiczZhuEstimator`):
ex15p-wedges-5

**Comment by tzanio** [comment]
> ex15p works for prisms (with ZienkiewiczZhuEstimator)

Super cool @jakubcerveny !

**Comment by jakubcerveny** [comment]
This should be ready for review.

**Comment by jakubcerveny** [comment]
This was a lot harder than I had initially thought. All the special cases that needed to be handled are due to the common configuration of a prism edge touching another prism/hex face.

**Comment by jakubcerveny** [comment]
I'm also archiving this picture here. This is the new constraint that had to be supported (edge constrained by a face). In the code it's handled as a degenerate face-face constraint (the point matrix rectangle has one zero dimension).

face-edge-constraint

**Comment by mlstowell** [comment]
I ran `ex6` on a twice-refined copy of `data\toroid-wedge.mesh` and the results look mostly good.
GLVis_s01
Notice that there is a slight clustering of small elements near the interfaces of the original coarse elements which are located at angles corresponding to multiples of pi/3.  I do not believe this is an artifact of the prism refinement strategy because I saw similar clustering when using `toroid-hex.mesh`.  I suspect it may be related to a lack of smoothness in the original geometry which uses an H1_3D_P3 approximation for the nodes.
The reason for choosing this mesh is that it is the only wedge mesh that I know of, amongst our standard meshes, which requires matching the triangular faces with non-trivial orientations.

BTW: This test cannot be done with the original `data\toroid-wedge.mesh` because it is only one element across and the error estimator only detects problems at element boundaries.  This leads to an extreme amount of clustering at the original element interfaces.

File: fem/fespace.hpp [hunk 1]
```diff
@@ -144,8 +144,9 @@ class FiniteElementSpace
    static inline int DecodeDof(int dof, double& sign)
    { return (dof >= 0) ? (sign = 1, dof) : (sign = -1, (-1 - dof)); }
 
-   /// Helper to get vertex, edge or face DOFs (entity=0,1,2 resp.).
+   /// Helpers to get vertex, edge or face DOFs (entity=0,1,2 resp.).
    void GetEntityDofs(int entity, int index, Array<int> &dofs) const;
+   void GetDegenerateFaceDofs(int index, Array<int> &dofs) const;
```
Comment (mlstowell) [inline fem/fespace.hpp]: I don't know what is meant by "degenerate face dofs".  A bit of explanation would be useful here or a reference to documentation elsewhere if that makes more sense.

Comment (jakubcerveny) [reply fem/fespace.hpp]: Yes, this one is tricky, added an explanation inside the method.

File: fem/pfespace.cpp [hunk 2]
```diff
@@ -147,20 +149,37 @@ void ParFiniteElementSpace::Construct()
...
       {
          ngedofs = pncmesh->GetNGhostEdges()
                    * fec->DofForGeometry(Geometry::SEGMENT);
       }
+
       if (pmesh->Dimension() > 2)
       {
-         ngfdofs = pncmesh->GetNGhostFaces()
-                   * fec->DofForGeometry(pncmesh->GetGhostFaceGeometry(0));
+         if (fdofs != NULL) // have mixed faces
```
Comment (mlstowell) [inline fem/pfespace.cpp]: I believe `fdofs` is allocated whenever face dofs exist in this space not only when the mesh has a mixture of face types.  You may need to use `Mesh::HasGeometry` to explicitly check for the presence of triangle and quadrilateral faces.

File: fem/pfespace.cpp [hunk 3]
```diff
@@ -1465,9 +1505,17 @@ void ParFiniteElementSpace::UnpackDof(int dof,
       dof -= nedofs;
       if (dof < nfdofs) // regular face
       {
-         MFEM_ASSERT(!pmesh->HasGeometry(Geometry::TRIANGLE), "");
-         int nf = fec->DofForGeometry(Geometry::SQUARE);
-         entity = 2, index = dof / nf, edof = dof % nf;
+         if (fdofs) // have mixed faces
```
Comment (mlstowell) [inline fem/pfespace.cpp]: Again, checking `fdofs` may not be sufficient.

File: mesh/ncmesh.cpp [hunk 4]
```diff
@@ -811,6 +853,13 @@ void NCMesh::RefineElement(int elem, char ref_type)
       return;
    }
 
+   /*std::cout << "Refining element " << elem << " ("
```
Comment (mlstowell) [inline mesh/ncmesh.cpp]: Perhaps this is no longer needed

**Comment by mlstowell** [comment]
I came up with another test mesh using two wedge elements sharing a quadrilateral face but rotated by 90 degrees.
GLVis_s01
This works in `ex6` which ran without error for 12 AMR iterations.  However, for some reason GLVis could not display the results after the ninth iteration.

The error from GLVis is the following:
```
Assertion failed: (eid.Size() > 0) is false:
 --> edge prism not found
 ... in function: void mfem::NCMesh::TraverseQuadFace(int, int, int, int, const mfem::NCMesh::PointMatrix &, int, mfem::NCMesh::Face **)
 ... in file: mesh/ncmesh.cpp:2300

Backtrace:
0) [0x10877ed2a]: mfem::mfem_error(char const*)
1) [0x108977771]: mfem::NCMesh::TraverseQuadFace(int, int, int, int, mfem::NCMesh::PointMatrix const&, int, mfem::NCMesh::Face**)
2) [0x10897654e]: mfem::NCMesh::TraverseQuadFace(int, int, int, int, mfem::NCMesh::PointMatrix const&, int, mfem::NCMesh::Face**)
3) [0x10897a6c1]: mfem::NCMesh::BuildFaceList()
4) [0x108c43a3e]: mfem::NCMesh::GetFaceList()
5) [0x108899ab3]: mfem::Mesh::GenerateNCFaceInfo()
6) [0x108899204]: mfem::Mesh::FinalizeTopology()
7) [0x1088a916d]: mfem::Mesh::Loader(std::__1::basic_istream<char, std::__1::char_traits<char> >&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)
8) [0x1088eda4c]: mfem::Mesh::Load(std::__1::basic_istream<char, std::__1::char_traits<char> >&, int, int, bool)
9) [0x1088a1e9c]: mfem::Mesh::Mesh(std::__1::basic_istream<char, std::__1::char_traits<char> >&, int, int, bool)
10) [0x1088a20c9]: mfem::Mesh::Mesh(std::__1::basic_istream<char, std::__1::char_traits<char> >&, int, int, bool)
11) [0x10869425c]: ReadStream(std::__1::basic_istream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
12) [0x1086ad505]: main
```
I'm not sure what to make of this but I will send @jakubcerveny the mesh files from each AMR iteration (they're fairly small).

File: fem/pfespace.cpp (hunk 2 above)
Comment (jakubcerveny) [reply fem/pfespace.cpp]: The comment may not be 100% precise, I can remove it. The code however should be OK, the reasoning is that if `fdofs != NULL`, it has to be used, no matter what the reason is behind it being non-NULL. That said, if all faces have the same geometry, it would be strange to have `fdofs` allocated, so maybe it's a bug elsewhere?

Comment (mlstowell) [reply fem/pfespace.cpp]: You make a good point.  Perhaps we need to take a closer look at `fdofs`.  Not necessarily as part of this PR but at some point.

File: mesh/ncmesh.cpp (hunk 4 above)
Comment (jakubcerveny) [reply mesh/ncmesh.cpp]: I would actually like to keep it for future debugging...

**Comment by jakubcerveny** [comment]
I was looking into `ex6` on @mlstowell's two-wedge test (above). Neither Mark nor me can reproduce the problem in `TraverseQuadFace` anymore. Not sure how that happened, let's hope it's gone.

There is however another problem with GLVis: after the first refinement in `ex6`, GLVis reports bad element orientations:
```
Elements with wrong orientation: 13 / 152 (NOT FIXED)

Assertion failed: (CheckElementOrientation(false) == 0) is false:
 --> 
 ... in function: virtual void mfem::Mesh::Finalize(bool, bool)
 ... in file: mesh/mesh.cpp:2102
```
These are some observations that I made:
* The mesh seems fine when saved and in my debug dumps, there is nothing suspicious about the elements.
* The number of elements with "wrong orientation" is random, it differs on every run.
* It only happens in GLVis.
* When MFEM+GLVis are compiled with MFEM_THREAD_SAFE, the problem disappears.

Finally, I ran GLVis under `helgrind` and this seems to be the problem:
```
==31193== Possible data race during write of size 8 at 0x9A15DC0 by thread #1
==31193== Locks held: 1, at address 0x54B30C8
==31193==    at 0x556604: mfem::Poly_1D::CalcChebyshev(int, double, double*) (fe.cpp:6828)
==31193==    by 0x57ADA3: mfem::Poly_1D::CalcBasis(int, double, double*) (fe.hpp:1756)
==31193==    by 0x55C11C: mfem::H1_TriangleElement::CalcShape(mfem::IntegrationPoint const&, mfem::Vector&) const (fe.cpp:7820)
==31193==    by 0x562612: mfem::H1_WedgeElement::CalcShape(mfem::IntegrationPoint const&, mfem::Vector&) const (fe.cpp:8622)
==31193==    by 0x60D801: mfem::IsoparametricTransformation::Transform(mfem::IntegrationRule const&, mfem::DenseMatrix&) (eltrans.cpp:487)
==31193==    by 0x4598C0: VisualizationSceneSolution3d::FindNewBox(bool) (vssolution3d.cpp:921)
==31193==    by 0x44A18B: VisualizationSceneScalarData::FindNewBoxAndValueRange(bool) (vsdata.hpp:146)
==31193==    by 0x441684: VisualizationSceneScalarData::DoAutoscale(bool) (vsdata.cpp:54)
==31193==    by 0x458F94: VisualizationSceneSolution3d::SetShading(int, bool) (vssolution3d.cpp:780)
==31193==    by 0x40B737: StartVisualization(int) (glvis.cpp:475)
==31193==    by 0x40F655: main (glvis.cpp:1567)
==31193== 
==31193== This conflicts with a previous write of size 8 by thread #2
==31193== Locks held: none
==31193==    at 0x556604: mfem::Poly_1D::CalcChebyshev(int, double, double*) (fe.cpp:6828)
==31193==    by 0x57ADA3: mfem::Poly_1D::CalcBasis(int, double, double*) (fe.hpp:1756)
==31193==    by 0x55C11C: mfem::H1_TriangleElement::CalcShape(mfem::IntegrationPoint const&, mfem::Vector&) const (fe.cpp:7820)
==31193==    by 0x562782: mfem::H1_WedgeElement::CalcDShape(mfem::IntegrationPoint const&, mfem::DenseMatrix&) const (fe.cpp:8643)
==31193==    by 0x60D3C2: mfem::IsoparametricTransformation::EvalJacobian() (eltrans.cpp:408)
==31193==    by 0x454DCE: mfem::ElementTransformation::Jacobian() (in /home/jakub/glvis/glvis)
==31193==    by 0x4AC038: mfem::Mesh::GetElementJacobian(int, mfem::DenseMatrix&) (mesh.cpp:63)
==31193==    by 0x4C07B6: mfem::Mesh::CheckElementOrientation(bool) (mesh.cpp:3771)
==31193==  Address 0x9a15dc0 is 0 bytes inside a block of size 16 alloc'd
==31193==    at 0x4C2FC2F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==31193==    by 0x414A3B: mfem::Memory<double>::New(int) (in /home/jakub/glvis/glvis)
==31193==    by 0x414A86: mfem::Memory<double>::New(int, mfem::MemoryType) (mem_manager.hpp:495)
==31193==    by 0x412B12: mfem::Vector::SetSize(int) (vector.hpp:416)
==31193==    by 0x55B7BD: mfem::H1_TriangleElement::H1_TriangleElement(int, int) (fe.cpp:7747)
==31193==    by 0x5619F2: mfem::H1_WedgeElement::H1_WedgeElement(int, int) (fe.cpp:8517)
==31193==    by 0x579868: __static_initialization_and_destruction_0(int, int) (fe.cpp:11968)
==31193==    by 0x5798B0: _GLOBAL__sub_I__ZN4mfem13FiniteElementC2EiNS_8Geometry4TypeEiii (fe.cpp:11974)
==31193==    by 0x6B20CC: __libc_csu_init (in /home/jakub/glvis/glvis)
==31193==    by 0x69187BE: (below main) (libc-start.c:247)
==31193==  Block was alloc'd by thread #1
```

So, it seems that the main thread of GLVis interferes with `Mesh::CheckElementOrientation`, which then reports spurious orientation errors. I'm not sure how to fix that, any idea @v-dobrev? Do we want to add some locks to GLVis?

**Comment by jakubcerveny** [comment]
With C++11, maybe we could simply make the shape Vectors `thread_local`.

**Comment by v-dobrev** [comment]
We can simply make the methods `CalcShape` and `CalcDShape` in class `H1_WedgeElement` thread safe -- basically always using the same code that is now used when `MFEM_THREAD_SAFE` is defined.

**Comment by jakubcerveny** [comment]
I tried that, but to make it work, the same needs to be done also in `H1_SegmentElement` and `H1_TriangleElement`, which is probably not what we want. 

The problem is that both threads use `mfem::WedgeFE` (defined in fe.cpp) and it has to be thread-safe all the way down. The only reason this problem doesn't show for other elements is that the globals used for `ElementTransformation` are special elements, e.g., `TriLinear3DFiniteElement HexahedronFE`. If the latter is changed to `H1_HexahedronElement`, I get the race condition too. 

The simplest solution seems to be to make `WedgeFE` `thread_local`, so that each thread has its own copy. I tried it, it seems to work. The question is whether we want to go that path. The initialization of these globals is already complicated (i.e., what order is the right one). 

But using `thread_local` would probably also make `MFEM_THREAD_SAFE` unneccessary, is that right? I feel that `thread_local` is the right solution, but it will be good to discuss it.

**Comment by jakubcerveny** [comment]
@v-dobrev, can we merge this? #949 depends on it and I'll soon have another branch that will need it...

**Comment by jakubcerveny** [comment]
Also, what do you think about making the global variable `WedgeFE` `thread_local`? To me it seems like a clean solution.

File: CHANGELOG [hunk 5]
```diff
@@ -53,7 +56,6 @@ GPU support
 
 Discretization improvements
 ---------------------------
-
```
Comment (v-dobrev) [inline CHANGELOG]: Please, revert this change.

**Comment by jakubcerveny** [comment]
After discussing with @v-dobrev a couple weeks ago I made the the global variable `WedgeFE` use the `thread_local` feature (new in C++11). It seems to be a nice and easy solution to the above GLVis race condition and I think we could use this feature more, even to deal with some things that are now under `MFEM_THREAD_SAFE`. The `thread_local` keyword is used indirectly through the new `MFEM_THREAD_LOCAL` macro that I added to `globals.hpp`.

**Comment by jakubcerveny** [comment]
Hm, it seems `thread_local` causes some trouble on Windows :-(  Not sure what the issue is

**Comment by jakubcerveny** [comment]
Removed `thread_local` from `WedgeFE` for the time being due to some problem on Windows. I will resolve this as a separate issue because this is in fact unrelated to `nc-prism-dev`.

**Comment by jakubcerveny** [comment]
This is ready, I'm gonna merge it in `next` so we see if there are any last problems.

**Comment by v-dobrev** [comment]
@jakubcerveny, when I build this branch with `-O3 -Wall` flags I see some warnings -- can you please take a look and resolve those.

**Comment by jakubcerveny** [comment]
The only warnings I saw were from `bilininteg_mass.cpp`, like this:
```
fem/bilininteg_mass.cpp:275:16: warning: ‘sol_x[7]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```
I fixed that by making `sol_x` `volatile`. Not sure this is what you wanted though, @v-dobrev.

**Comment by v-dobrev** [comment]
The warnings I see are in `mesh/ncmesh.*` -- this is with older gcc (4.8.5) -- I'll take a look.

I have not seen the warnings you fixed -- what compiler did you see these with? Also, I'm not sure using `volotile` is a good choice there -- it will prevent various optimizations. @camierjs, what do you think?

**Comment by v-dobrev** [comment]
Follow up on the warnings in `mesh/ncmesh.*` -- these seem to be limited to older versions of gcc (4.8.5) because versions 4.9.x and later do not produce the same warnings. So I think we can leave this as it is.

**Comment by jakubcerveny** [comment]
I have a pretty old compiler too: 5.4.0. 

Maybe I should remove the `volatile`s -- they might do more harm than good (?)

**Comment by camierjs** [comment]
We can remove the `volatile` qualifier, the issue seems to be more compiler related.

**Comment by v-dobrev** [comment]
I think the current state is fine -- I do not see any warnings with gcc 7 and gcc 8 when building with `-O3 -pedantic -Wall`.

**Comment by v-dobrev** [comment]
The more extensive internal tests uncovered an issue with the following sample run:
```console
mpirun -np 4 ./ex15p -m ../data/square-disc-nurbs.mesh -no-vis
```
I did some debugging and it looks like the issue is that on at least on processor, the variable `nmat` in the method https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/fespace.cpp#L1166-L1190
has a value of `0` which does not seem right. A value of `0` may be OK for mixed meshes but this is not a mixed mesh.

**Comment by v-dobrev** [comment]
I added a few range checks in class `DenseTensor` which catch the above issue.

**Comment by jakubcerveny** [comment]
Thanks, that was a major bug -- sorry. The checks in `DenseTensor` are useful.

**Comment by v-dobrev** [comment]
Thanks for resolving this so quickly!
